### PR TITLE
MSA retirement is free for valid requests

### LINKED ISSUE #340: MSA Retirement should be Free
## Related
- https://github.com/LibertyDSNP/frequency/issues/44

Should be able to follow the pattern from https://github.com/LibertyDSNP/frequency/pull/323 

As a user, I want to be able to retire my MSA, even if I have _zero_ tokens.

## Acceptance Criteria
- [x] Given that an MSA has **one** key, I should be able to retire my MSA without any tokens.

## Notes
- #44
- #205
- #229 
- https://github.com/LibertyDSNP/frequency/pull/323

### PR DESCRIPTION
**Pull request by rlaferla**
Implemented free MSA retirement for valid requests Modified unit test `test_retire_msa_success()` to check that valid requests don't pay Added unit test `test_retire_msa_invalid_pays()` to check that invalid requests pay

Closes #340

### CODE REVIEW

File: pallets/msa/src/lib.rs [hunk 1]
```diff
@@ -579,6 +579,11 @@ pub mod pallet {
 
 		/// Retire a MSA
 		///
+		/// When a user wants to disassociate themselves from Frequency, they can retire their MSA for free provided that:
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:582]: Perhaps, we can extend the description to say that retiring your MSA does not altogether remove all associated data. For example, messages related to an MSA are not removed.

File: pallets/msa/src/lib.rs [hunk 2]
```diff
@@ -614,9 +620,10 @@ pub mod pallet {
 			Self::delete_key_for_msa(msa_id, &who)?;
 			Self::deposit_event(Event::KeyRemoved { key: who });
 
-			// Deposit the "MsaRetired" event
 			Self::deposit_event(Event::MsaRetired { msa_id });
-			Ok(())
+
+			// MSA retirement is free
+			Ok(Pays::No.into())
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:626]: you can remove this

File: pallets/msa/src/lib.rs [hunk 3]
```diff
@@ -589,12 +594,13 @@ pub mod pallet {
 		/// - Returns [`RegisteredProviderCannotBeRetired`](Error::RegisteredProviderCannotBeRetired) if the MSA id is a registered provider
 
 		#[pallet::weight((T::WeightInfo::retire_msa(), DispatchClass::Normal, Pays::Yes))]
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:596]: add Pays::No here.

File: pallets/msa/src/lib.rs (hunk 2 above)
Comment (enddynayn) [reply pallets/msa/src/lib.rs:626]: Ahh. Perhaps I am confused.

In this case, we are charging a fee when it errors. I am not sure if this is intentional.

File: pallets/msa/src/lib.rs (hunk 1 above)
Comment (rlaferla) [reply pallets/msa/src/lib.rs:580]: The intent is for it to remove it but we are not doing that currently.  I will update the description to explain what it is doing currently.

Comment (rlaferla) [reply pallets/msa/src/lib.rs:580]: /// This does not currently remove any messages related to the MSA.

File: pallets/msa/src/lib.rs (hunk 2 above)
Comment (rlaferla) [reply pallets/msa/src/lib.rs:626]: Yes, it's conditional.  It's free for valid requests but we charge a fee for invalid requests.

Comment (enddynayn) [reply pallets/msa/src/lib.rs:626]: I am not opposed to this idea. However, we do not do this with any other free transaction. Can you please share with me why we decided to do it this way for retiring?

File: pallets/msa/src/tests.rs [hunk 4]
```diff
@@ -15,12 +15,12 @@ use common_primitives::{
 };
 use common_runtime::extensions::check_nonce::CheckNonce;
 use frame_support::{
-	assert_err, assert_noop, assert_ok,
+	assert_err, assert_err_with_weight, assert_noop, assert_ok,
```
Comment (enddynayn) [inline pallets/msa/src/tests.rs:18]: I did not know about this one `assert_err_with_weight` but it looks like we are not using it.

File: pallets/msa/src/lib.rs (hunk 2 above)
Comment (rlaferla) [reply pallets/msa/src/lib.rs:626]: I have redesigned it to use the signed extension.

File: pallets/msa/src/lib.rs (hunk 3 above)
Comment (rlaferla) [reply pallets/msa/src/lib.rs]: Removed

File: pallets/msa/src/tests.rs (hunk 4 above)
Comment (rlaferla) [reply pallets/msa/src/tests.rs:18]: Removed

**Pull request closed without merging**